src/combinatorics/simple_graph/basic.lean

```diff
@@ -313,6 +317,72 @@ instance decidable_mem_edge_set [decidable_rel G.adj] :
 instance edges_fintype [decidable_eq V] [fintype V] [decidable_rel G.adj] :
   fintype G.edge_set := subtype.fintype _
 
+/-! ## Darts -/
+
+/-- A `dart` is an oriented edge, implemented as an ordered pair of adjacent vertices. -/
+@[ext, derive decidable_eq]
+structure dart extends V × V :=
+(is_adj : G.adj fst snd)
+
+section darts
+variables {G}
+
+instance dart.fintype [fintype V] [decidable_rel G.adj] : fintype G.dart :=
+fintype.of_equiv (Σ v, G.neighbor_set v)
+{ to_fun := λ s, ⟨(s.fst, s.snd), s.snd.property⟩,
+  inv_fun := λ d, ⟨d.fst, d.snd, d.is_adj⟩,
+  left_inv := λ s, by ext; simp,
+  right_inv := λ d, by ext; simp }
+
+/-- The edge associated to the dart. -/
+def dart.edge (d : G.dart) : sym2 V := ⟦d.to_prod⟧
+
+@[simp] lemma dart.edge_mem (d : G.dart) : d.edge ∈ G.edge_set :=
+d.is_adj
+
+/-- The dart with reversed orientation from a given dart. -/
+def dart.rev (d : G.dart) : G.dart :=
+⟨d.to_prod.swap, G.symm d.is_adj⟩
+
+@[simp] lemma dart.rev_edge : Π (d : G.dart), d.rev.edge = d.edge :=
+by { rintro ⟨⟨v,w⟩, _⟩, exact sym2.eq_swap }
+
+@[simp] lemma dart.rev_rev (d : G.dart) : d.rev.rev = d :=
+by ext; refl
```

Comment:
```suggestion
@[simp] lemma dart.rev_rev (d : G.dart) : d.rev.rev = d := dart.ext _ _ $ prod.swap_swap _
```

Reply:
I'm not sure it's worth our time doing this, but, well, it does elaborate faster (1.6ms vs 12ms, such speed!), so I guess I better not waste your effort.